lib/staccato/v4/tracker.rb

```diff
@@ -1,155 +1,334 @@
+# frozen_string_literal: true
+
+require_relative 'adapter_defaults'
+require_relative 'event' # must load first, to dynamically define event methods
+
 module Staccato::V4
   # The `Tracker` class has methods to create all `Hit` types
-  #   using the tracker and client id
+  #   using the measurement_id, api_secret, and client_id.
   #
   # @author Tony Pitale
   class Tracker
-    attr_accessor :events
-
-    attr_reader :measurement_id, :api_secret
+    extend Staccato::V4::AdapterDefaults
 
     # sets up a new tracker
-    # @param measurement_id [String] the measurement id from GA
-    # @param api_secret [String] the required api secret key
-    # @param client_id [String, nil] unique value to track user sessions
-    # @param options [Hash]
-    def initialize(measurement_id, api_secret, client_id = nil, options = {})
-      @measurement_id = measurement_id
-      @api_secret = api_secret
-      @client_id = client_id
-      @adapters = []
+    # @param measurement_id [String] the identifier for a Data Stream
+    #     provided by google, i.e., `G-XXXXXXXX`
+    # @param api_secret [String] the required API Secret that is generated
+    #     through the Google Analytics UI
+    # @param client_id [String] uniquely identifies a user instance of a web
+    #     client
+    # @param user_id [String, nil] A unique identifier for a user
+    # @param user_properties [Hash] custom user properties that describe
+    #     segments of your user base, such as language preference or geographic
+    #     location
+    # @param non_personalized_ads [Boolean] Set to true to indicate these events
+    #     should not be used for personalized ads
+    # @param time [Time] The time to associate with the event.  This should
+    #     only be set to record events that happened in the past.  This value
+    #     can be overridden via user_property or event timestamps.
+    # @param default_event_params [Hash] default event params used by all events
+    #     sent from this tracker
+    def initialize(measurement_id, api_secret, client_id,
+                   user_id: nil, user_properties: nil,
+                   non_personalized_ads: nil,
+                   time: nil, timestamp_micros: nil,
+                   default_event_params: {})
+      raise ArgumentError, 'measurement_id is required' unless measurement_id
+      raise ArgumentError, 'api_secret is required'     unless api_secret
+      raise ArgumentError, 'client_id is required'      unless client_id
```

Comment:
I think letting client_id be a random UUID as it does in V3 is acceptable.

Reply:
Yeah, I puzzled over this one.

Requiring `client_id` is a big change, but I believe it is necessary for to truly satisfy Google's v4 API.  As far as I can tell, there is no documentation anywhere for using anything other than the `gtag` or `gtm` generated `client_id`.  The big push to v4 _seems_ to be at motivated by regulatory compliance concerns.  And my interpretation of the docs is that access to standard analytics `client_id` _could_ potentially be a matter of end-user consent (depending on what you're tracking).

In other words, I believe it might be wrong for the library to automatically generate a missing `client_id`.  If users of the gem think that it is reasonable and permissible to generate a missing client_id, they should manage that themselves.

Also, using a randomly generated UUID can result in tremendously inflated session counts and your boss telling you to try something else.  Ask me how I know.  😉

Reply:
I'm not sure how a random UUID would be a matter of end-user consent.

I could see it being a session issue, as it's going to make a new session for each post.

Is there a way we could split the difference? Maybe make it a configuration somewhere? `require_client_id: true` by default, and a user could disable that with false to get the random UUID back?

Reply:
> I'm not sure how a random UUID would be a matter of end-user consent.

Being forced get the real `client_id` also forces the developer into a position where it should be easy for them to also check for consent.  That _might_ encourage the developer to _consider_ whether they're allowed to send everything they send.

Not knowing the real `client_id` suggests that we haven't even attempted to check whether we have end-user consent.  We might have both on the frontend but never communicated it to the backend.  Or we might not have a `client_id` _because_ we don't have consent. 
 Either way, that doesn't mean we can't send _any_ data, but (as I understand it) it does mean we shouldn't send certain private or user identifying data.  It might be that you're fine sending everything.  In my case, we will send certain `user_properties` and custom parameters and custom events if we have consent, and we won't if we don't.  It might be that you are fine sending _everything_ you intended to send.

I'd prefer a config setting to just doing it by default.  But personally, I think it's better to force the developer to make their own decision about how to handle it:
* Google's docs never give any valid alternatives to the official `client_id`.
* Unique client IDs for each tracker instance causes an _explosion_ in session count.
* The app probably has internal session identifiers of its own that are more stable.

IMO it would be fine to _document_ that `SecureRandom.uuid` seems to more-or-less work.  But, assuming you're using rails, the following would work much better and we'd be better off recommending something like it:

```ruby
class Current < ActiveSupport::CurrentAttributes
  attribute :ga_real_cid
  attribute :ga_fake_cid
  attribute :app_session_id

  def ga_cid = ga_real_cid || self.ga_fake_cid ||= app_session_id || SecureRandom.uuid

end
```

...using `before_filters` in `ApplicationController` to get and set these values appropriately.  They could even be forwarded into background jobs and used there as well, using `Current.set(ga_real_cid:, ga_fake_cid:, app_session_id:) { the bg job code }`.

Reply:
Thanks for all the info. As you know I haven't been working with this for awhile, so this is all very useful. I appreciate the time you've taken to explain it.

Okay, as long as we document what they'll have to do in case they just need to get it working ASAP, I'm fine moving forward with this being required.

Reply:
@tpitale thanks!  and there's some other bits of this PR that could use more docs too.  not least of which is adding something to the README.md.  😉